composite/apy-finance/src/registry/index.ts

```diff
@@ -1,39 +1,62 @@
-import { ethers } from 'ethers'
-import registryAbi from './IRegistry.json'
-import assetAllocationAbi from './IAssetAllocation.json'
+import { ethers, utils } from 'ethers'
+import registryAbi from '../abi/IRegistry.json'
+import assetAllocationAbi from '../abi/IAssetAllocation.json'
+import erc20Abi from '../abi/ERC20.json'
 
-const getRegistry = (address: string, rpcUrl: string, abi: any): ethers.Contract => {
-  const provider = new ethers.providers.JsonRpcProvider(rpcUrl)
-  return new ethers.Contract(address, abi, provider)
+type GetDecimals = (address: string) => Promise<number>
+type GetTokenDecimals = (provider: ethers.providers.JsonRpcProvider, abi: any) => GetDecimals
+
+const getTokenDecimals: GetTokenDecimals = (provider, abi) => async (address) => {
+  return await new ethers.Contract(address, abi, provider).decimals()
 }
 
 type Allocations = {
   components: string[]
   units: string[]
 }
 
-export type GetAllocations = (registry: ethers.Contract) => () => Promise<Allocations>
+export type GetAllocations = (
+  registry: ethers.Contract,
+  getDecimals: GetDecimals,
+) => () => Promise<Allocations>
 
-const getAllocations: GetAllocations = (registry) => async () => {
+const getAllocations: GetAllocations = (registry, getDecimals) => async () => {
   const tokenAddresses = await registry.getTokenAddresses()
-  const [components, units]: string[][] = await Promise.all([
+  const [components, units, decimals]: string[][] = await Promise.all([
     Promise.all(tokenAddresses.map((address: string) => registry.symbolOf(address))),
-    Promise.all(tokenAddresses.map((address: string) => registry.balanceOf(address))),
+    Promise.all(
+      tokenAddresses.map(async (address: string) =>
+        utils.bigNumberify(await registry.balanceOf(address)),
+      ),
+    ),
+    Promise.all(tokenAddresses.map(async (address: string) => getDecimals(address))),
   ])
-  return { components, units }
+
+  return {
+    components,
+    units: units.map((unit, i) => {
+      const decimal = new utils.BigNumber(10).pow(decimals[i])
+      return new utils.BigNumber(unit).div(decimal).toString()
```

Comment:
This is not correct. It behaves differently than DeFi Pulse.

You either need to return `units` + `decimals`, or normalize `units` as the number with 18 decimals. Like they do in the DeFi Pulse `getAllocations` function.

Reply:
I'd say this is correct. It's true that is missing the default value for `decimal` to be `1e18`, but the calculation is correct.

Reply:
I don't agree.

In this PR you changed that the token-allocation adapter accepts decimal numbers on input and not base units. This is not a good API. These are big numbers and we should work with base units.

Then this apy-finance adapter returns a decimal number. In that sense it's correct.

But then defi-pulse adapter returns a int number of units (normalized to 18 decimal points), and when you pipe that to token-allocation adapter which now expects decimal numbers, you are going to get a wrong result.

Please calculate everything in units, just print the formatted decimal price on the output.

Reply:
Also looks like ethers `BigNumber` does not support decimal numbers, so your `.div` is going to round the result. Take a look at [this comment](https://github.com/ethers-io/ethers.js/issues/452#issuecomment-472168526) on how to do it properly.